src/meta/Intersection.ts

```diff
@@ -0,0 +1,167 @@
+import global from '@dojo/shim/global';
+import WeakMap from '@dojo/shim/WeakMap';
+import { Base } from './Base';
+
+interface IntersectionDetail {
+	entries: WeakMap<Element, IntersectionObserverEntry>;
+	intersectionObserver?: IntersectionObserver; // attached observer
+	options: IntersectionGetOptions;
+	root: string;
+	rootMargin: string | undefined;
+	thresholds: number[]; // thresholds the observe should be attached with
+}
+
+export interface IntersectionGetOptions {
+	root?: string;
+	rootMargin?: string;
+	threshold?: number;
+	thresholds?: number[];
+}
+
+export interface IntersectionResult {
+	intersectionRatio: number;
+	isIntersecting: boolean;
+}
+
+const defaultIntersection: IntersectionResult = Object.freeze({
+	intersectionRatio: 0,
+	isIntersecting: false
+});
+
+export class Intersection extends Base {
+	private _details: IntersectionDetail[] = [];
+
+	private _getDetails(options: IntersectionGetOptions): IntersectionDetail {
+		let {
+			root = '',
+			rootMargin,
+			threshold,
+			thresholds = []
+		} = options;
+		const {
+			_details: details
+		} = this;
+		if (typeof threshold === 'number' && !thresholds.length) {
+			thresholds = [ threshold ];
+		}
+
+		// Look to see if a detail record has already been created for these options
+		let cached: IntersectionDetail | undefined = undefined;
+		for (const detail of details) {
+			if (
```

Comment:
I think we can simplify this if condition some what....

Reply:
In terms of formatting or?

Reply:
The logic is pretty complicated, i'd either break these logical conditions up so each one is done individually, like:

```ts
if (options !== detail.options) {
    continue;
}
```

Or perhaps even better, use a different mechanism for the options caching (like a map, using the serialised options as the key)? It would look something like this:

```ts
	private _details: Map<string, IntersectionDetail> = new Map();

	private _getDetails(options: IntersectionGetOptions): IntersectionDetail {
		const cacheKey = JSON.stringify(options);
		let cached = this._details.get(cacheKey);

		if (!cached) {
			const entries = new WeakMap<Element, IntersectionObserverEntry>();
			cached = { ...options, entries };
			this._details.set(cacheKey, cached);
		}
		return cached;
	}
```